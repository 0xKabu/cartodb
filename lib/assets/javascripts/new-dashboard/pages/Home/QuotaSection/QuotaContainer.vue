<template>
  <div class="quota-container">
    <div class="quota-headers">
      <div class="quota-cell cell--main">
        <h3 class="title is-caption quota-title">{{title}}</h3>
      </div>

      <div class="quota-labels">
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey">{{ $t('QuotaSection.header.remaining') }}</span>
        </div>
        <div class="quota-cell cell--medium cell--mobile">
          <span class="text is-small is-txtSoftGrey">{{ $t('QuotaSection.header.used') }}</span>
        </div>
        <div class="quota-cell cell--medium">
          <span class="text is-small is-txtSoftGrey quota-total" :class="{'is-active' : perMonth }">{{$t(`QuotaSection.header.total`)}}</span>
          <span class="text is-small is-txtSoftGrey quota-month" :class="{'is-active' : perMonth }">&nbsp;{{$t(`QuotaSection.header.perMonth`)}}</span>
        </div>
        <div class="quota-cell cell--small"></div>
      </div>
    </div>
    <div class="quota-info">
      <slot></slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'QuotaContainer',
  props: {
    title: String,
    perMonth: Boolean
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quota-info {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  background-color: $white;
}

.quota-title {
  margin-bottom: 8px;
}

.quota-headers {
  display: flex;
  justify-content: space-between;
}

.quota-labels {
  display: flex;
}

.quota-cell {
  display: flex;
  position: relative;
  flex-grow: 0;
  flex-shrink: 0;
  align-items: center;
  padding: 0 10px;
  text-overflow: ellipsis;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }

  @media (max-width: $layout-tablet) {
    padding: 0 5px;
  }
}

.cell--mobile {
  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.cell--main {
  flex-grow: 1;
  flex-shrink: 1;
  min-width: 300px;

  @media (max-width: $layout-mobile) {
    min-width: unset;
  }
}

.cell--medium {
  width: 120px;

  @media (max-width: $layout-tablet) {
    width: 100px;
  }
}

.cell--small {
  width: 58px;

  @media (max-width: $layout-tablet) {
    display: none;
  }
}

.quota-month {
  display: none;
}

.is-active {
  &.quota-total {
    @media (max-width: $layout-tablet) and (min-width: $layout-mobile) {
      &::after {
        content: '*';
      }
    }
  }

  &.quota-month {
    display: flex;

    @media (max-width: $layout-tablet) and (min-width: $layout-mobile) {
      display: none;
    }

    &::after {
      content: '*';
    }
  }
}

</style>
