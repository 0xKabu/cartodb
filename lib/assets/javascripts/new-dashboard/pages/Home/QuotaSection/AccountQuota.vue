<template>
  <QuotaContainer :title="title" :perMonth="perMonth">
    <QuotaWidget :name="$t(`QuotaSection.storage`)"
      :usedQuota="getAmountInUnit(usedStorage, amountExponent)"
      :availableQuota="getAmountInUnit(availableStorage, amountExponent)"
      :unit="getUnit(amountExponent)"
      :formatToLocale="false"
      :helpLink="storageHelpLink"
      ></QuotaWidget>

    <QuotaWidget v-if="hasTableLimits" :name="$t(`QuotaSection.publicMaps`)"
      :usedQuota="usedPublicMaps"
      :availableQuota="availablePublicMaps"
      helpLink=""
      ></QuotaWidget>
    <QuotaWidget v-if="hasPublicMapLimits" :name="$t(`QuotaSection.datasets`)"
      :usedQuota="usedDatasets"
      :availableQuota="availableDatasets"
      helpLink=""
      ></QuotaWidget>
    <QuotaWidget v-if="hasApiKeysLimits" :name="$t(`QuotaSection.apiKeys`)"
      :usedQuota="usedApiKeys"
      :availableQuota="availableApiKeys"
      helpLink=""
      ></QuotaWidget>
  </QuotaContainer>
</template>

<script>
import { mapState } from 'vuex';
import QuotaWidget from './QuotaWidget';
import QuotaContainer from './QuotaContainer';
import CartoNode from 'carto-node';
import apiKeysCollectionTypes from 'dashboard/data/api-keys-collection-types';

const limitsUsers = ['Professional', '[DEDICATED]'];

export default {
  name: 'AccountQuota',
  components: {
    QuotaWidget,
    QuotaContainer
  },
  props: {
    title: String,
    perMonth: Boolean
  },
  beforeMount () {
    this.getApiKeysTotal();
  },
  data () {
    return {
      client: new CartoNode.AuthenticatedClient(),
      usedApiKeys: 0
    };
  },
  computed: {
    ...mapState({
      availableStorage: state => state.user.quota_in_bytes,
      remainingStorage: state => state.user.remaining_byte_quota,
      availablePublicMaps: state => state.user.public_map_quota,
      linkMapsTotal: state => state.user.link_privacy_map_count,
      passwordMapsTotal: state => state.user.password_privacy_map_count,
      publicMapsTotal: state => state.user.public_privacy_map_count,
      availableDatasets: state => state.user.table_quota,
      usedDatasets: state => state.user.table_count,
      availableApiKeys: state => state.user.regular_api_key_quota,
      planAccountType: state => state.user.account_type
    }),
    usedStorage () {
      return this.availableStorage - this.remainingStorage;
    },
    amountExponent () {
      return this.getExpBaseTwo(this.availableStorage) - 10;
    },
    storageHelpLink () {
      return 'https://carto.com/help/your-account/your-disk-storage/';
    },
    hasTableLimits() {
      return limitsUsers.includes(this.planAccountType);
    },
    hasPublicMapLimits() {
      return limitsUsers.includes(this.planAccountType);
    },
    hasApiKeysLimits() {
      return limitsUsers.includes(this.planAccountType);
    },
    usedPublicMaps () {
      return this.linkMapsTotal + this.passwordMapsTotal + this.publicMapsTotal;
    }
  },
  methods: {
    getExpBaseTwo (sizeInBytes) {
      if (sizeInBytes === 0) {
        return 0;
      }

      let exponent = 0;
      if (Math.log2) {
        exponent = Math.log2(sizeInBytes);
      } else {
        exponent = Math.log(sizeInBytes) * Math.LOG2E;
      }

      return Math.round(exponent / 10) * 10;
    },
    getUnit (exponent) {
      if (exponent < 10) {
        return 'B';
      } else if (exponent < 20) {
        return 'KB';
      } else if (exponent < 30) {
        return 'MB';
      } else if (exponent < 40) {
        return 'GB';
      } else if (exponent < 50) {
        return 'TB';
      } else if (exponent < 60) {
        return 'PB';
      } else if (exponent < 70) {
        return 'EB';
      } else {
        return '?';
      }
    },
    getAmountInUnit (number, exponent) {
      return number / Math.pow(2, exponent);
    },
    getApiKeysTotal () {
      this.client.getApiKeys(apiKeysCollectionTypes.REGULAR,

        (err, _, data) => {
          if (err) {
            return;
          }
          this.usedApiKeys = data.count;
        }
      );
    }
  }
}
</script>
