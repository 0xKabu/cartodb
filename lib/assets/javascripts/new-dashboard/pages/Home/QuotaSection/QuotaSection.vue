<template>
  <div class="section is-bgSoftBlue">
    <div class="container">
      <div class="full-width">
        <SectionTitle class="grid-cell" :title="$t(`QuotaSection.title`)">
          <template slot="icon">
            <img svg-inline src="../../../assets/icons/section-title/quota.svg">
          </template>
        </SectionTitle>
        <ul class="grid quota-list">
          <li class="grid-cell grid-cell--col12 quota-listitem">
            <QuotaContainer :title="$t(`QuotaSection.disk`)" :perMonth=false>
              <QuotaWidget :name="$t(`QuotaSection.storage`)" :usedQuota="divideBaseTwo(usedStorage, getBaseTwo)" :availableQuota="divideBaseTwo(availableStorage, getBaseTwo)" :unit="getUnitFromBaseTwo(getBaseTwo)"></QuotaWidget>
            </QuotaContainer>
          </li>
          <li class="grid-cell grid-cell--col12 quota-listitem">
            <QuotaContainer :title="$t(`QuotaSection.dataServices`)" :perMonth=true>
              <QuotaWidget :name="$t(`QuotaSection.geocoding`)" :usedQuota="geocodingUsed" :availableQuota="geocodingAvailable"></QuotaWidget>
              <QuotaWidget :name="$t(`QuotaSection.isolines`)" :usedQuota="isolinesUsed" :availableQuota="isolinesAvailable"></QuotaWidget>
              <QuotaWidget :name="$t(`QuotaSection.routing`)" :usedQuota="routingUsed" :availableQuota="routingAvailable"></QuotaWidget>
            </QuotaContainer>
          </li>
        </ul>
        <div class="quota-billing">
          <span class="quota-billingday text is-small is-txtSoftGrey">{{ $t(`QuotaSection.credits`, { day: billingDay })}}</span>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import QuotaWidget from './QuotaWidget';
import QuotaContainer from './QuotaContainer';
import SectionTitle from 'new-dashboard/components/SectionTitle';
import format from 'date-fns/format';

export default {
  name: 'QuotasModule',
  components: {
    QuotaWidget,
    QuotaContainer,
    SectionTitle
  },
  computed: {
    ...mapState({
      availableStorage: state => state.user.quota_in_bytes,
      remainingStorage: state => state.user.remaining_byte_quota,
      geocodingUsed: state => state.user.geocoding.monthly_use,
      geocodingAvailable: state => state.user.geocoding.quota,
      routingUsed: state => state.user.mapzen_routing.monthly_use,
      routingAvailable: state => state.user.mapzen_routing.quota ? state.user.mapzen_routing.quota : 0,
      isolinesUsed: state => state.user.here_isolines.monthly_use,
      isolinesAvailable: state => state.user.here_isolines.quota,
      billingPeriod: state => state.user.next_billing_period
    }),
    usedStorage () {
      return this.availableStorage - this.remainingStorage;
    },
    getBaseTwo () {
      return this.baseTwoRepresentation(this.availableStorage);
    },
    billingDay () {
      return format(new Date(this.billingPeriod), 'Do');
    }
  },
  methods: {
    baseTwoRepresentation (initialByte) {
      let i = 0;
      let value = initialByte;
      while (value > 1024) {
        value = value / 1024;
        i = i + 1;
      }
      return (i * 10);
    },
    getUnitFromBaseTwo (baseTwo) {
      if (baseTwo < 10) {
        return 'B';
      } else if (baseTwo < 20) {
        return 'Kb';
      } else if (baseTwo < 30) {
        return 'MB';
      } else if (baseTwo < 40) {
        return 'GB';
      } else if (baseTwo < 50) {
        return 'TB';
      } else if (baseTwo < 60) {
        return 'PB';
      } else if (baseTwo < 70) {
        return 'EB';
      } else {
        return '?';
      }
    },
    divideBaseTwo (number, baseTwo) {
      return number / Math.pow(2, baseTwo);
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quota-listitem {
  &:not(:first-of-type) {
    margin-top: 64px;
  }
}

.quota-billing {
  display: flex;
  justify-content: flex-end;
  padding: 0 10px;
}

.quota-billingday {
  margin-top: 16px;

  &::before {
    content: '*';
  }
}
</style>
