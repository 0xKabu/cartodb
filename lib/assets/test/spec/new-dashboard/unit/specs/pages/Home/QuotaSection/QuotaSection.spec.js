import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import usersArray from '../../../fixtures/users';
import QuotaSection from 'new-dashboard/pages/Home/QuotaSection/QuotaSection';

const localVue = createLocalVue();
localVue.use(Vuex);

let store = new Vuex.Store({
  state: {
    user: usersArray[2]
  }
});

const $t = key => key;

describe('QuotaSection.vue', () => {
  it('should render correct contents', () => {
    const quotaSection = shallowMount(QuotaSection, {
      store,
      localVue,
      mocks: {
        $t
      }
    });

    expect(quotaSection).toMatchSnapshot();
  });

  describe('Computed', () => {
    it('should return all the correct computed variables', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let availableStoarage = quotaSection.vm.availableStorage;
      let remainingStorage = quotaSection.vm.remainingStorage;
      let geocodingUsed = quotaSection.vm.geocodingUsed;
      let geocodingAvailable = quotaSection.vm.geocodingAvailable;
      let routingUsed = quotaSection.vm.routingUsed;
      let routingAvailable = quotaSection.vm.routingAvailable;
      let isolinesUsed = quotaSection.vm.isolinesUsed;
      let isolinesAvailable = quotaSection.vm.isolinesAvailable;
      let usedStorage = quotaSection.vm.usedStorage;
      let getBaseTwo = quotaSection.vm.getBaseTwo;

      expect(availableStoarage).toBe(1073741824);
      expect(remainingStorage).toBe(1066631168);
      expect(geocodingUsed).toBe(103);
      expect(geocodingAvailable).toBe(1200);
      expect(routingUsed).toBe(54);
      expect(routingAvailable).toBe(102);
      expect(isolinesUsed).toBe(134);
      expect(isolinesAvailable).toBe(1332);
      expect(usedStorage).toBe(7110656);
      expect(getBaseTwo).toBe(20);
    });

    it('should return correct unit if you give a base two', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotaSection.vm.getUnitFromBaseTwo(0);
      let result2 = quotaSection.vm.getUnitFromBaseTwo(12);
      let result3 = quotaSection.vm.getUnitFromBaseTwo(20);
      let result4 = quotaSection.vm.getUnitFromBaseTwo(35);
      let result5 = quotaSection.vm.getUnitFromBaseTwo(49);
      let result6 = quotaSection.vm.getUnitFromBaseTwo(51);
      let result7 = quotaSection.vm.getUnitFromBaseTwo(62);
      let result8 = quotaSection.vm.getUnitFromBaseTwo(70);

      expect(result1).toBe('B');
      expect(result2).toBe('Kb');
      expect(result3).toBe('MB');
      expect(result4).toBe('GB');
      expect(result5).toBe('TB');
      expect(result6).toBe('PB');
      expect(result7).toBe('EB');
      expect(result8).toBe('?');
    });

    it('should divide the number passed by the base two provided', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotaSection.vm.divideBaseTwo(0, 0);
      let result2 = quotaSection.vm.divideBaseTwo(10, 0);
      let result3 = quotaSection.vm.divideBaseTwo(2097152, 20);
      let result4 = quotaSection.vm.divideBaseTwo(132070244352, 30);

      expect(result1).toBe(0);
      expect(result2).toBe(10);
      expect(result3).toBe(2);
      expect(result4).toBe(123);
    });
  });

  describe('Methods', () => {
    it('should return correct base two of the number (ten by ten)', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotaSection.vm.baseTwoRepresentation(0);
      let result2 = quotaSection.vm.baseTwoRepresentation(1395864371);
      let result3 = quotaSection.vm.baseTwoRepresentation(1363148);

      expect(result1).toBe(0);
      expect(result2).toBe(30);
      expect(result3).toBe(20);
    });

    it('should return correct unit if you give a base two', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotaSection.vm.getUnitFromBaseTwo(0);
      let result2 = quotaSection.vm.getUnitFromBaseTwo(32);
      let result3 = quotaSection.vm.getUnitFromBaseTwo(12);

      expect(result1).toBe('B');
      expect(result2).toBe('GB');
      expect(result3).toBe('Kb');
    });

    it('should divide the number passed by the base two provided', () => {
      const quotaSection = shallowMount(QuotaSection, {
        store,
        localVue,
        mocks: {
          $t
        }
      });

      let result1 = quotaSection.vm.divideBaseTwo(0, 0);
      let result2 = quotaSection.vm.divideBaseTwo(10, 0);
      let result3 = quotaSection.vm.divideBaseTwo(2097152, 20);
      let result4 = quotaSection.vm.divideBaseTwo(132070244352, 30);

      expect(result1).toBe(0);
      expect(result2).toBe(10);
      expect(result3).toBe(2);
      expect(result4).toBe(123);
    });
  });
});
