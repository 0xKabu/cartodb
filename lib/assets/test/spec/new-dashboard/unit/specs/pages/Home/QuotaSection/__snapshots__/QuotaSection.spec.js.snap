// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaSection.vue should render correct contents 1`] = `
<div class="section is-bgSoftBlue section--noBorder">
  <div class="container">
    <div class="full-width">
      <sectiontitle-stub title="QuotaSection.title" class="grid-cell"><template></template></sectiontitle-stub>
      <ul class="grid quota-list">
        <li class="grid-cell grid-cell--col12 quota-listitem">
          <accountquota-stub title="QuotaSection.account"></accountquota-stub>
        </li>
        <li class="grid-cell grid-cell--col12 quota-listitem">
          <dataservicesquota-stub title="QuotaSection.dataServices" permonth="true"></dataservicesquota-stub>
        </li>
      </ul>
      <div class="quota-billing"><span class="quota-billingday text is-small is-txtSoftGrey">
          QuotaSection.credits
        </span></div>
    </div>
  </div>
</div>
`;
